go-controller/pkg/util/multi_network.go

```diff
@@ -213,6 +215,124 @@ func (layer3NetConfInfo *Layer3NetConfInfo) MTU() int {
+		err = fmt.Errorf("new %s netconf subnets %v has changed, expect %v",
+			types.Layer2Topology, newLayer2NetConfInfo.subnets, layer2NetConfInfo.subnets)
+		errs = append(errs, err)
+	}
+	if layer2NetConfInfo.mtu != newLayer2NetConfInfo.mtu {
+		err = fmt.Errorf("new %s netconf MTU %v has changed, expect %v",
+			types.Layer2Topology, newLayer2NetConfInfo.mtu, layer2NetConfInfo.mtu)
+		errs = append(errs, err)
+	}
+	if !reflect.DeepEqual(layer2NetConfInfo.excludeCIDRs, newLayer2NetConfInfo.excludeCIDRs) {
+		err = fmt.Errorf("new %s netconf ExcludeCIDRs %v has changed, expect %v",
+			types.Layer2Topology, newLayer2NetConfInfo.excludeCIDRs, layer2NetConfInfo.excludeCIDRs)
+		errs = append(errs, err)
+	}
+	if len(errs) != 0 {
+		err = kerrors.NewAggregate(errs)
+		klog.V(5).Infof(err.Error())
+		return false
+	}
+	return true
+}
+
+func newLayer2NetConfInfo(netconf *ovncnitypes.NetConf) (*Layer2NetConfInfo, error) {
+	clusterSubnets, excludeIPs, err := verifyExcludeIPs(netconf.Subnets, netconf.ExcludeCIDRs)
+	if err != nil {
+		return nil, fmt.Errorf("invalid %s netconf %s: %v", netconf.Topology, netconf.Name, err)
+	}
+
+	return &Layer2NetConfInfo{
+		subnets:        netconf.Subnets,
+		mtu:            netconf.MTU,
+		excludeCIDRs:   netconf.ExcludeCIDRs,
+		ClusterSubnets: clusterSubnets,
+		ExcludeIPs:     excludeIPs,
+	}, nil
+}
+
+// parseL2ClusterSubnets parses and returns the subnets in the layer2 netconf
+func parseL2ClusterSubnets(clusterSubnetString string) ([]*net.IPNet, error) {
+	var subnetList []*net.IPNet
+	subnetStringList := strings.Split(clusterSubnetString, ",")
+
+	for _, subnetString := range subnetStringList {
+		_, subnet, err := net.ParseCIDR(subnetString)
+		if err != nil {
+			return nil, err
+		}
+
+		subnetList = append(subnetList, subnet)
+	}
+
+	if len(subnetList) == 0 {
+		return nil, fmt.Errorf("failed to parse any CIDRs from %q", clusterSubnetString)
+	}
+
+	return subnetList, nil
+}
+
+func verifyExcludeIPs(netCidr string, excludeCIDRs []string) ([]*net.IPNet, []net.IP, error) {
+	clusterSubnets, err := parseL2ClusterSubnets(netCidr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cluster subnet %s is invalid: %v", netCidr, err)
+	}
+
+	excludeIPs := make([]net.IP, 0)
+	for _, excludeCIDRstr := range excludeCIDRs {
+		_, excludeCIDR, err := net.ParseCIDR(excludeCIDRstr)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid subnet %q provided in the exclude_cidrs list %v: %v",
+				excludeCIDRstr, excludeCIDRs, err)
+		}
+
+		for excludeIP := excludeCIDR.IP; excludeCIDR.Contains(excludeIP); excludeIP = NextIP(excludeIP) {
+			found := false
+			for _, subnet := range clusterSubnets {
+				if subnet.Contains(excludeIP) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return nil, nil, fmt.Errorf("ip to be excluded %q is not part of any of the provided Network CIDRs (%v)",
+					excludeIP, clusterSubnets)
+			}
+			excludeIPs = append(excludeIPs, excludeIP)
+		}
```

Comment:
This could be done more efficiently I guess.

Would this work? https://stackoverflow.com/questions/40406230/net-ipnet-inside-other-net-ipnet

Reply:
but we'd need each specific excludeIP anyway, to be used by reserveIP later.

Reply:
Well, that's true. But my guess is that the real effort would be to add support to the allocator to allocate ranges. If we deem that too much effort, at least we could iterate the IPs some place closer to the allocator itself. If everywhere else we just require to know how if subnets overlap or not, we can do that in a more optimum way in those places. And when we  eventually add support for the allocator to allocate ranges, then we would only have to change the allocater itself and nothing else.

Reply:
but validation should be done when NAD is added, right? If we move it later, we'd only find configuration error when controller/allocator is created.

Reply:
So I guess the whole concern I have is iterating all IPs of a subnet in multiple places in the code (now here and also in the L2 controller Init function). An IPv6 network can be very big. The allocator itself should be able to efficiently block ranges with whatever math it needs to happen. If we are not able to handle that now, let's keep it specific to the allocator, that single place. But here we could just do things more efficiently.

Reply:
> but validation should be done when NAD is added, right? If we move it later, we'd only find configuration error when controller/allocator is created.

I am not saying that you don't validate here. Just do it in a more efficient way if possible, not iterating through all IPs. That stack overflow article I linked has a suggestion on how to do that.

Reply:
But then we'd need to iterate each excluded-ips later when allocate IP. So it only delay the inefficiency later, right?

Reply:
@jcaamano I've made the change to do validation the light way. But we'd still iterate each excluded-ips after allocator is initialized. Please take a look